pkg/controller/drain.go

```diff
@@ -487,85 +493,112 @@ func filterSharedPVs(pvMap map[string][]string) {
+func (o *DrainOptions) evictPodsWithPv(attemptEvict bool, pods []*corev1.Pod,
 	policyGroupVersion string,
 	getPodFn func(namespace, name string) (*api.Pod, error),
 	returnCh chan error,
-	stopCh <-chan struct{},
 ) {
-	remainingPods := 0
-
 	sortPodsByPriority(pods)
 
 	volMap := o.doAccountingOfPvs(pods)
 
-EvictingPods:
-	for {
-		retryPods := []*corev1.Pod{}
-
-		for i, pod := range pods {
-			select {
-			case <-stopCh:
-				glog.Warningf("The caller function returned. No need to try evictions now")
-				return
-			default:
+	var (
+		remainingPods []*api.Pod
+		fastTrack     bool
+		nretries      = int(o.MaxEvictRetries)
+	)
+
+	if attemptEvict {
+		for i := 0; i < nretries; i++ {
+			remainingPods, fastTrack = o.evictPodsWithPVInternal(attemptEvict, pods, volMap, policyGroupVersion, getPodFn, returnCh)
+			if fastTrack || len(remainingPods) == 0 {
+				//Either all pods got evicted or we need to fast track the return (node deletion detected)
+				break
 			}
 
-			err := o.evictPod(*pod, policyGroupVersion)
+			glog.V(4).Info("Eviction for some pods will be retried after ", PodEvictionRetryInterval)
+			pods = remainingPods
+			remainingPods = nil
+			time.Sleep(PodEvictionRetryInterval)
+		}
 
-			if apierrors.IsTooManyRequests(err) {
-				// Pod eviction failed because of PDB violation, we will retry one we are done with this list.
-				glog.V(3).Info("Pod ", pod.Namespace, "/", pod.Name, " couldn't be evicted because of PDB violation. Will be retried.")
-				retryPods = append(retryPods, pod)
-				continue
-			} else if apierrors.IsNotFound(err) {
-				glog.V(3).Info("\t", pod.Name, " is already gone")
+		if fastTrack {
+			// Fast tract the return (node deletion detected)
+			for range remainingPods {
+				glog.V(4).Info("Returning success for remaining pods")
+				// This is executed when node is not found anymore.
+				// Return success to caller for all non-processed pods so that the caller function can move on.
 				returnCh <- nil
-				continue
-			} else if err != nil {
-				returnCh <- fmt.Errorf("Error when evicting pod: %v/%v. Will not be retried. Err: %v", pod.Namespace, pod.Name, err)
-				continue
 			}
+		}
 
-			glog.V(3).Info("waiting for PVs to detach from node for pod: ", pod.Name)
-			// Eviction was successful. Wait for pvs for this pod to detach
-			pvs := volMap[pod.Namespace+"/"+pod.Name]
-			err = o.waitForDetach(pvs, o.nodeName, stopCh)
+		if len(remainingPods) > 0 {
+			// Force delete the pods remaining after evict retries. No pods are expected to survive this and errors
+			// are returned via returnCh anyway.
+			_, _ = o.evictPodsWithPVInternal(false, pods, volMap, policyGroupVersion, getPodFn, returnCh)
+		}
+	} else {
+		// Force delete the pods remaining after evict retries. No pods are expected to survive this and errors
+		// are returned via returnCh anyway.
+		_, _ = o.evictPodsWithPVInternal(false, pods, volMap, policyGroupVersion, getPodFn, returnCh)
```

Comment:
Yes. But we cannot keep retrying can we?

Reply:
If delete fails, shouldn't we fail the drain, and subsequently fail the machine delete operation? This is mainly because of azure case in which the azure driver detaches the disk, but kcm tries to re-attach the disk. Error is not sent via `returnCh` if delete fails with this change.